File: game_plugins/hey_danke_fuer_den_fisch/shared/sc/plugin2015/Board.java
package sc.plugin2015;

import java.util.LinkedList;
import java.util.List;

import com.thoughtworks.xstream.annotations.XStreamAlias;

import sc.plugin2015.util.Constants;

/**
 * Klasse welche eine Spielbrett darstellt. Bestehend aus einem
 * zweidimensionalen Array aus Feldern
 * 
 * @author soed
 * 
 */
@XStreamAlias(value = "board")
public class Board implements Cloneable {

	private Field[][] fields;

	/**
	 * Konstruktor, der ein zufällig generiertes Spielfeld erzeugt.
	 */
	public Board() {
		this.init();
	}

	/**
	 * Konstruktor, der entweder ein zufällig generiertes Spielfeld oder ein
	 * leeres erzeugt. Nur für den Server relevant.
	 * 
	 * @param init
	 *            Zufallsgeneration an/aus
	 */
	public Board(boolean init) {
		if (init)
			this.init();
		else
			this.makeClearBoard();
	}

	/**
	 * generiert ein neues, leeres Spielfeld.
	 */
	private void makeClearBoard() {
		fields = new Field[Constants.ROWS][Constants.COLUMNS];
		for (int x = 0; x < Constants.COLUMNS; x++) {
			for (int y = 0; y < Constants.ROWS; y++) {
				fields[x][y] = new Field();
			}
		}
	}

	/**
	 * Generiert ein neues Spielfeld mit zufällig auf dem Spielbrett verteilten
	 * Fischen.
	 */
	private void init() {

		fields = new Field[Constants.ROWS][Constants.COLUMNS];
		int rnd;

		List<Integer> fish = new LinkedList<Integer>();

		for (int i = 0; i < 3; i++) {
			for (int j = 0; j < Constants.FISH[i]; j++) {
				fish.add(i + 1);
			}
		}
		for (int x = 0; x < Constants.COLUMNS; x++) {
			for (int y = 0; y < Constants.ROWS; y++) {
				if ((y & 1) == 0 && x == Constants.COLUMNS - 1) {
					fields[x][y] = new Field();
				} else {
					rnd = (int) (Math.random() * fish.size());
					fields[x][y] = new Field(fish.get(rnd));
					fish.remove(rnd);
				}
			}
		}
	}

	/**
	 * Gibt das Feld mit angegebenen Koordinaten zurück
	 * 
	 * @param x x-Koordinate
	 * @param y y-Koordinate
	 * @return das Feld an den Koordinaten
	 */
	public Field getField(int x, int y) {
		return fields[x][y];
	}

	/**
	 * Gibt alle Felder im Board als zweidimensionales Array zurück
	 * @return alle Felder
	 */
	public Field[][] getBoard() {
		return this.fields;
	}

	/**
	 * Liefert den Pinguin auf den Koordinaten
	 * @param x x-Koordinate
	 * @param y y-Koordinate
	 * @return Pinguin auf dem Feld mit den Koordinaten
	 */
	public Penguin getPenguin(int x, int y) {
		return fields[x][y].getPenguin();
	}

	/**
	 * Setzt einen Pinguin an gewählte Koordinaten. Diese Methode ist nur für
	 * den Server relevant, da keine vollständige Überprüfung auf korrekte Züge
	 * durchgeführt wird. Um für einen Spieler einen neuen Pinguin zu setzen,
	 * die {@link sc.plugin2015.SetMove#perform(GameState, Player) perform}
	 * -Methode benutzen
	 * 
	 * @param x
	 *            X-Koordinate
	 * @param y
	 *            Y-Koordinate
	 * @param penguin
	 *            Pinguin, der gesetzt werden soll.
	 */
	public void putPenguin(int x, int y, Penguin penguin)
			throws IllegalArgumentException {
		if (x < 0 || y < 0 || x >= Constants.COLUMNS || y >= Constants.ROWS
				|| fields[x][y].getFish() != 1
				|| fields[x][y].getPenguin() != null)
			throw new IllegalArgumentException();
		this.fields[x][y].putPenguin(penguin);
	}

	/**
	 * nur für den Server relevant
	 */
	private void putPenguinMove(int x, int y, Penguin penguin)
			throws IllegalArgumentException {
		if (x < 0 || y < 0 || x >= Constants.COLUMNS || y >= Constants.ROWS
				|| fields[x][y].getFish() <= 0
				|| fields[x][y].getPenguin() != null)
			throw new IllegalArgumentException();
		this.fields[x][y].putPenguin(penguin);
	}

	/**
	 * Gibt zurück, ob sich ein Pinguin auf dem Feld befindet
	 * 
	 * @param x x-Koordinate
	 * @param y y-Koordinate
	 * @return true falls sich ein Pinguin auf dem Feld befindet
	 */
	public boolean penguin(int x, int y) {
		if (fields[x][y].getPenguin() == null)
			return false;
		else
			return true;
	}

	/**
	 * Gibt zurück, ob ein Spieler an einer Position einen Pinguin hat.
	 * 
	 * @param x x-Koordinate
	 * @param y y-Koordinate
	 * @param color
	 *            Die Spielerfarbe
	 * @return true wenn sich ein Pinguin des Spielers an der Position befindet
	 */
	public boolean hasPinguin(int x, int y, PlayerColor color) {
		if (fields[x][y].getPenguin() == null)
			return false;
		if (fields[x][y].getPenguin().getOwner() == color)
			return true;
		else
			return false;

	}

	/**
	 * Gibt die Anzahl der Fische an bestimmten Koordinaten zurück
	 * 
	 * @param x x-Koordinate
	 * @param y y-Koordinate
	 * 
	 * @return integer Anzahl der Fische
	 * 
	 */
	public int getFishNumber(int x, int y) {
		return fields[x][y].getFish();
	}

	/**
	 * Bewegt einen Pinguin von einem Startfeld auf ein Zielfeld. Diese Methode
	 * ist nur für den Server relevant, da hier keine Fehlerüberprüfung
	 * durchgeführt wird. Zum Ausführen von Zügen die
	 * {@link sc.plugin2015.RunMove#perform(GameState, Player) perform}-Methode
	 * benutzen.
	 * 
	 * @param fromX
	 * @param fromY
	 *            das Startfeld auf dem sich der Pinguin befindet
	 * @param toX
	 * @param toY
	 *            das Zielfeld auf das der Pinguin bewegt werden soll
	 * @param color
	 *            die Farbe des Besitzers
	 */
	public void movePenguin(int fromX, int fromY, int toX, int toY,
			PlayerColor color) {
		Penguin penguin = fields[fromX][fromY].removePenguin(color);
		fields[fromX][fromY].setFish(0);
		putPenguinMove(toX, toY, penguin);
	}

	/**
	 * Gibt eine deep copy des Objektes zurück. Von allen Objekten, welche diese
	 * Klasse beherbergt werden auch Kopien erstellt.
	 * 
	 * @see java.lang.Object#clone()
	 */
	@Override
	public Object clone() throws CloneNotSupportedException {
		Board clone = new Board(false);
		for (int x = 0; x < Constants.COLUMNS; x++) {
			for (int y = 0; y < Constants.ROWS; y++) {
				clone.fields[x][y].fish = this.getField(x, y).getFish();
				if (this.getField(x, y).hasPenguin())
					clone.fields[x][y].putPenguin(new Penguin(this.getPenguin(
							x, y).getOwner()));
				// else
				// clone.fields[x][y].putPenguin(null);
			}
		}
		return clone;

	}

	public boolean equals(Object o) {
		if (!(o instanceof Board))
			return false;
		Board board = (Board) o;

		for (int x = 0; x < Constants.COLUMNS; x++) {
			for (int y = 0; y < Constants.ROWS; y++) {
				if (!(this.getPenguin(x, y) == null && board.getPenguin(x, y) == null)
						&& (!(this.getPenguin(x, y).equals(
								board.getPenguin(x, y)) || !(this
								.getFishNumber(x, y) == board.getFishNumber(x,
								y)))))
					return false;
				}
		}
		return true;
	}
}

File: plugin/src/main/kotlin/sc/plugin2023/Board.kt
package sc.plugin2023

import com.thoughtworks.xstream.annotations.XStreamAlias
import sc.api.plugins.*
import sc.api.plugins.Coordinates
import kotlin.random.Random
import sc.plugin2023.util.PenguinConstants as Constants

/**
 * Klasse welche eine Spielbrett darstellt. Bestehend aus einem
 * zweidimensionalen Array aus Feldern
 *
 * @author soed
 */
@XStreamAlias(value = "board")
class Board(override val gameField: MutableTwoDBoard<Field> = generateFields()): RectangularBoard<Field>(gameField) {
    
    constructor(board: Board): this(board.gameField.deepCopy())
    
    override fun isValid(coordinates: Coordinates) =
            (coordinates.x + coordinates.y) % 2 == 0 &&
            coordinates.x >= 0 &&
            super.isValid(coordinates.copy(coordinates.x / 2))
    
    /** Prüft, ob auf dieser [position] bereits eine Spielfigur ist. */
    fun isOccupied(position: Coordinates): Boolean =
            this[position].isOccupied
    
    /** Gibt das Feld an den gegebenen Koordinaten zurück. */
    override operator fun get(x: Int, y: Int) =
            super.get(x / 2, y)
    
    /** Ersetzt die Fische des Feldes durch einen Pinguin.
     * @return Anzahl der ersetzten Fische. */
    operator fun set(position: Coordinates, team: Team?): Int {
        if(!isValid(position))
            outOfBounds(position)
        val field = gameField[position.y][position.x / 2]
        gameField[position.y][position.x / 2] = Field(penguin = team)
        return field.fish
    }
    
    fun possibleMovesFrom(pos: Coordinates) =
            HexDirection.values().flatMap { direction ->
                (1 until Constants.BOARD_SIZE).map {
                    Move.run(pos, direction.vector * it)
                }.takeWhile { getOrEmpty(it.to).fish > 0 }
            }
    
    /** Returns a list of the non-null filter outputs */
    fun <T> filterFields(filter: (Field, Coordinates) -> T?): Collection<T> =
            gameField.flatMapIndexed { y, row ->
                row.mapIndexedNotNull { x, field ->
                    filter(field, Coordinates(x * 2 + y % 2, y))
                }
            }
    
    fun getPenguins() =
            filterFields { field, coordinates ->
                field.penguin?.let { Pair(coordinates, it) }
            }
    
    fun getOrEmpty(key: Coordinates?) = key?.let { getOrNull(it) } ?: Field()
    
    override val entries: Set<Map.Entry<Coordinates, Field>>
        get() = filterFields { f, coordinates -> Positioned(coordinates, f) }.toSet()
    
    override fun clone(): Board = Board(this)
    
    companion object {
        /** Generiert ein neues Spielfeld mit zufällig auf dem Spielbrett verteilten Fischen. */
        private fun generateFields(seed: Int = Random.nextInt()): MutableTwoDBoard<Field> {
            var remainingFish = Constants.BOARD_SIZE * Constants.BOARD_SIZE
            val random = Random(seed)
            println("Board Seed: $seed")
            var maxholes = 5
            // Pro Hälfte 32 Felder, mind. 27 Schollen
            // Maximal (64-20)/2 = 22 2-Fisch-Schollen,
            // also immer mindestens 5 1-Fisch-Schollen pro Seite
            return Array(Constants.BOARD_SIZE / 2) {
                Array(Constants.BOARD_SIZE) {
                    val rand = random.nextInt(remainingFish)
                    if(rand < maxholes) {
                        maxholes--
                        Field()
                    } else {
                        val fish = (rand - maxholes) / 20 + 1
                        remainingFish -= fish
                        Field(fish)
                    }
                }
            }.let {
                it + it.reversedArray().map { list ->
                    Array(Constants.BOARD_SIZE) { index -> list[Constants.BOARD_SIZE - index - 1].clone() }
                }
            }
        }
        
    }
}
